src/utilities/filetypes/AlfalfaJSON.cpp

```diff
@@ -0,0 +1,223 @@
+        root[point.id().get()] = point.toJSON();
+      }
+      return root;
+    }
+  }  // namespace detail
+
+  AlfalfaJSON::AlfalfaJSON() : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl())) {}
+
+  AlfalfaJSON::AlfalfaJSON(const std::string& s) : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl(s))) {}
+
+  AlfalfaJSON::AlfalfaJSON(const openstudio::path& p) : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl(p))) {}
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeConstant(float value, const std::string& display_name) {
+    AlfalfaConstant component(value);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeMeter(const std::string& meter_name, const std::string& display_name) {
+    AlfalfaMeter component(meter_name);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeActuator(const std::string& component_name, const std::string& component_type,
+                                                            const std::string& control_type, const std::string& display_name) {
+    AlfalfaActuator component(component_name, component_type, control_type);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeOutputVariable(const std::string& variable_key, const std::string& variable_name,
+                                                                  const std::string& display_name) {
+    AlfalfaOutputVariable component(variable_key, variable_name);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeGlobalVariable(const std::string& variable_name, const std::string& display_name) {
+    AlfalfaGlobalVariable component(variable_name);
+    return exposeFromComponent(component, display_name);
+  }
+
+  void AlfalfaJSON::exposePoint(const AlfalfaPoint& point) {
+    m_impl->exposePoint(point);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeFromObject(const openstudio::IdfObject& idf_object, const std::string& display_name) {
+    IddObjectType idd_type = idf_object.iddObject().type();
+    std::string _display_name = display_name;
+
+    if (display_name.size() == 0) {
+      boost::optional<std::string> name = getName(idf_object);
+      if (name.is_initialized()) {
+        _display_name = name.get();
+      }
+    }
+
+    AlfalfaComponent* component;
+    try {
+      switch (idd_type.value()) {
+        case IddObjectType::OS_Output_Meter:
+        case IddObjectType::Output_Meter:
+          component = new AlfalfaMeter(idf_object);
+          break;
+        case IddObjectType::OS_EnergyManagementSystem_Actuator:
+        case IddObjectType::EnergyManagementSystem_Actuator:
+          component = new AlfalfaActuator(idf_object);
+          break;
```

Comment:
I'm noting that you're comparing the iddObjectType once here, then a second time in the Derived AlfalfaComponent's ctor.

While I'm sure this can work, I'm just going to say I dislike it because:
* Doing work twice, but that is probably plently fine, as the slowdown isn't likely to be noticeable
* Mostly because that duplicates the code. Say AlfalfaMeter tomorrow needs to support Output:Meter:CustomDecrement, we have to go modify two files.

Reply:
Yeah, this is the tension I always run up against. I considered making different constructors for the different model object types, but you still need to do idd_type checking to prevent bad arguments since idfObjects don't have separate c++ types. You could completely remove this function, but given that this method will probably only run 100ish times per workflow, I doubt the performance impact will be noticeable at all.